messaginginapp/src/main/java/io/customer/messaginginapp/state/InAppMessagingManager.kt

```diff
@@ -0,0 +1,85 @@
+package io.customer.messaginginapp.state
+
+import io.customer.messaginginapp.gist.presentation.GistListener
+import io.customer.sdk.core.di.SDKComponent
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+import org.reduxkotlin.Store
+import org.reduxkotlin.applyMiddleware
+import org.reduxkotlin.threadsafe.createThreadSafeStore
+
+data class InAppMessagingManager(val listener: GistListener? = null) {
+    private val store: Store<InAppMessagingState> = createStore()
+    internal val storeStatFlow = MutableStateFlow(store.state)
+    internal val scope: CoroutineScope = SDKComponent.scopeProvider.lifecycleListenerScope
+
+    init {
+        store.subscribe {
+            storeStatFlow.value = store.state
+        }
+    }
+
+    private fun createStore(): Store<InAppMessagingState> {
+        return createThreadSafeStore(
+            reducer = inAppMessagingReducer,
+            preloadedState = InAppMessagingState(),
+            applyMiddleware()
+        )
+    }
+
+    fun dispatch(action: InAppMessagingAction) = store.dispatch(action)
+
+    fun subscribe(listener: (InAppMessagingState) -> Unit): () -> Unit {
+        return store.subscribe { listener(store.state) }
+    }
+
+    fun getCurrentState() = store.state
+
+    fun <T> subscribeToAttribute(
+        selector: (InAppMessagingState) -> T,
+        areEquivalent: (old: T, new: T) -> Boolean = { old, new -> old == new },
+        listener: (T) -> Unit
+    ): Job {
+        return scope.launch {
+            storeStatFlow
+                .map(selector)
+                .distinctUntilChanged(areEquivalent)
+                .collect {
+                    listener(it)
+                }
+        }
+    }
+}
+
+internal fun <T1, T2> InAppMessagingManager.subscribeToAttributes(
+    selector1: (InAppMessagingState) -> T1,
+    selector2: (InAppMessagingState) -> T2,
+    areEquivalent: (old: Pair<T1, T2>, new: Pair<T1, T2>) -> Boolean = { old, new -> old == new },
+    listener: (T1, T2) -> Unit
+): Job {
+    return scope.launch {
+        storeStatFlow
+            .map { Pair(selector1(it), selector2(it)) }
+            .distinctUntilChanged(areEquivalent)
+            .collect { (v1, v2) -> listener(v1, v2) }
+    }
+}
+
+internal fun <T1, T2, T3> InAppMessagingManager.subscribeToAttributes(
```

Comment:
They method don't work the same as the original method, for multiple values, we need multiple sector, experience with list was not ideal. 

Secondly, 
```
areEquivalent: (old: InAppMessagingState, new: InAppMessagingState) -> Boolean = { old, new -> old == new },

areEquivalent: (old: Pair<T1, T2>, new: Pair<T1, T2>) -> Boolean = { old, new -> old == new },
```

are different, as we are just equating on attributes of states, not the complete state.

Reply:
I think what you are referring to serves a different purpose. Instead of selecting/observing attributes, you are observing states. with this, we will need always to provide `areEquivalent` for any attribute we want to observe. 

I am okay with adding that as replacement of subscribing to multiple attributes.

Reply:
Awesome 👍🏻